### ISSUE TITLE
custom @wp.func_grad that needs more shared memory than its forward

### ISSUE THREAD

**Issue by etaoxing** [NVIDIA/warp#1646]
### Bug Description

```python
"""A custom @wp.func_grad that needs more shared memory than its forward (here a
tile_matmul vs an elementwise op) makes Warp under-size the backward kernel's shared
memory -> out-of-bounds __shared__ writes on CUDA (error 700; CPU tolerates it). Fix:
fold custom_grad_func.get_total_required_shared() into the kernel roofline in
codegen.py::Adjoint.add_call (and rm -rf ~/.cache/warp/* to drop stale kernels)."""

import numpy as np

import warp as wp

wp.init()
M = wp.constant(32)


@wp.func
def scale2x(x: wp.array2d(dtype=float), y: wp.array2d(dtype=float), i: int):
    # forward: y = 2x elementwise -> tiny shared footprint
    wp.tile_store(y, wp.tile_load(x, shape=(M, M), offset=(i * M, 0)) * 2.0, offset=(i * M, 0))


@wp.func_grad(scale2x)
def adj_scale2x(x: wp.array2d(dtype=float), y: wp.array2d(dtype=float), i: int):
    # backward is dL/dx = 2*dL/dy, but routed through a shared-heavy tile_matmul (the
    # trigger); *0.0 keeps the value at 2 while the matmul still feeds the scatter (no DCE)
    g = wp.tile_load(wp.adjoint[y], shape=(M, M), offset=(i * M, 0))
    x0 = wp.tile_load(x, shape=(M, M), offset=(i * M, 0))
    wp.tile_atomic_add(wp.adjoint[x], g * 2.0 + wp.tile_matmul(g, x0) * 0.0, offset=(i * M, 0))


@wp.kernel(module="unique")
def run(x: wp.array2d(dtype=float), y: wp.array2d(dtype=float)):
    scale2x(x, y, wp.tid())


def main(dev="cuda:0"):
    m, b = int(M), 128
    x = wp.array(np.ones((b, m), np.float32), device=dev, requires_grad=True)
    y = wp.zeros_like(x)
    with wp.Tape() as tape:
        wp.launch_tiled(run, dim=(b // m,), inputs=[x], outputs=[y], block_dim=64, device=dev)
    tape.backward(grads={y: wp.ones_like(y)})
    print("no OOB; x.grad all == 2 ?", np.allclose(x.grad.numpy(), 2.0))  # CUDA 700 here without the fix


if __name__ == "__main__":
    main()
```

### System Information

```bash
Software
  Warp (package):   1.15.0
  Warp (core):      1.15.0
  warp-clang:       1.15.0 (LLVM 18.1.3)
  NumPy:            2.5.1
  Python:           3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0]
  Platform:         Linux-6.17.0-35-generic-x86_64-with-glibc2.39

CUDA
  Enabled:          True
  Toolkit:          12.9
  Driver:           13.2
  NVRTC:            12.9
  Forward compat:   True

Libraries
  MathDx:           0.3.1
  cuBQL:            enabled
  NanoVDB:          32.8.0

Build
  Compiler:         GCC 14.2.1
  Debug:            False
  Verify FP:        False
  Fast math:        False
  Sanitizer:        none

CPU
  Model:            arrowlake-s
  ISA features:     61
                      64bit, adx, aes, avx, avx2, avxifma, avxneconvert, avxvnni, ...

Devices
  cpu
    Name:             x86_64
  cuda:0
    Name:             NVIDIA GeForce RTX 5090
    Memory:           31.4 GiB
    Arch:             sm_120
    SMs:              170
    PCI:              00000000:02:00
    Mempool:          enabled
    GPU->CPU mem:     True
    CPU->GPU mem:     False
    CPU/GPU atomics:  False
````

**Comment by shi-eric** [comment]
This is addressed by 69e82cda5ee576dc95b5fbd60e011907b5d281d0.

The change sizes custom `@wp.func_grad` and `@wp.func_replay`
shared-memory frames separately for backward launches. This prevents a
backward kernel from under-reserving shared memory when the custom
gradient or replay needs more scratch than the forward function, without
inflating the forward launch reservation.

Public API behavior:

- No new API or signature changes were introduced.
- Existing custom gradients and replays using shared-memory tile
  operations now receive sufficient backward shared memory, while
  forward kernels reserve only their forward requirements.

Test coverage:

- `warp/tests/tile/test_tile_shared_memory.py` adds
  `test_tile_custom_grad_extra_shared` and
  `test_tile_custom_grad_shared_forward`, covering correct gradients and
  separate forward/backward reservations.
- `warp/tests/test_grad_customs.py` adds
  `test_custom_grad_helper_backward_propagation` and
  `test_backward_use_propagation_across_kernels`, covering build-order
  independence across shared helpers.
- `warp/tests/test_ref.py` adds
  `test_ref_shared_helper_backward_raises`, covering deferred validation
  independent of kernel build order.

The issue's original 32×32 `tile_matmul` reproducer also completed on
CUDA without an illegal memory access and produced the expected gradient
of 2 throughout.
